proposals/xxxx-nested-protocols.md

```diff
@@ -0,0 +1,150 @@
+
+As an additional benefit, within the context of `TableView`, the nested protocol `Delegate` can be referred to by a shorter name (as is the case with all other nested types):
+
+```swift
+class TableView {
+  weak var delegate: Delegate?
+  
+  protocol Delegate { /* ... */ }
+}
+```
+
+Protocols can also be nested within non-generic functions and closures. Admittedly, this is of somewhat limited utility, as all conformances to such protocols must also be within the same function. However, there is also no reason to artificially limit the complexity of the models which developers create within a function. Some codebases (of note, the Swift compiler itself) make use of large closures with nested types, and they beneift from abstractions using protocols.
+
+```swift
+func doSomething() {
+
+   protocol Abstraction {
+     associatedtype ResultType
+     func requirement() -> ResultType
+   }
+   struct SomeConformance: Abstraction {
+     func requirement() -> Int { ... }
+   }
+   struct AnotherConformance: Abstraction {
+     func requirement() -> String { ... }
+   }
+   
+   func impl<T: Abstraction>(_ input: T) -> T.ResultType {
+     // ...
+   }
+   
+   let _: Int = impl(SomeConformance())
+   let _: String = impl(AnotherConformance())
+}
+```
+
+## Detailed design
+
+Protocols may be nested anywhere that a struct/class/enum/actor may be nested, with the exception of generic contexts. For example, the following remains forbidden:
+
+```swift
+class TableView<Element> {
+
+  protocol Delegate {  // Error: protocol 'Delegate' cannot be nested within a generic context.
+    func didSelect(_: Element)
+  }
+}
+```
+
+The same applies to generic functions:
+
+```swift
+func genericFunc<T>(_: T) {
+  protocol Abstraction {  // Error: protocol 'Abstraction' cannot be nested within a generic context.
+  }
+}
+```
+
+And to other functions within generic contexts:
+
+```swift
+class TableView<Element> {
+  func doSomething() {
+    protocol MyProtocol {  // Error: protocol 'Abstraction' cannot be nested within a generic context.
+    }
+  }
+}
+```
+
+Supporting this would require either:
+
+- Introducing generic protocols, or
+- Mapping generic parameters to associated types.
+
+Neither is in in-scope for this proposal, but this author feels there is enough benefit here even without supporting generic contexts. Either of these would certainly make for interesting future directions.
+
+## Source compatibility
+
+This feature is additive.
+
+## ABI compatibility
+
+This proposal is purely an extension of the language's ABI and does not change any existing features.
+
+As with other nested types, the parent context forms part of the mangled name of a nested protocol. Therefore, moving the protocol around is an ABI-incompatible change.
+
+## Implications on adoption
+
+This feature can be freely adopted and un-adopted in source
+code with no deployment constraints and without affecting source or ABI
+compatibility.
```

Comment:
You speak to adoption constraints in the above section:

>As with other nested types, the parent context forms part of the mangled name of a nested protocol. Therefore, moving the protocol around is an ABI-incompatible change.

I think this sentence belongs in **Implications on adoption**. You might also consider mentioning that moving an existing protocol inside a type is, of course, a source-breaking change, but for non-resilient libraries, you can make it a source-compatible change using a type-alias to the newly-nested protocol.